src/Passes/Source/Validator/Validator.hpp

```diff
@@ -0,0 +1,86 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "AllocationManager/AllocationManager.hpp"
+#include "AllocationManager/IAllocationManager.hpp"
+#include "ValidationPass/ValidationConfiguration.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+#include <memory>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    /// Validator class that defines a set of rules which constitutes the profile definition. Each of
+    /// the rules can be used to transform a generic QIR and/or validate that the QIR is compliant with
+    /// said rule.
+    class Validator
+    {
+      public:
+        using ValidatorPtr = std::unique_ptr<Validator>;
+
+        // Constructors
+        //
+
+        explicit Validator(
+            ValidationPassConfiguration const& cfg,
+            bool                               debug,
+            llvm::TargetMachine*               target_machine = nullptr);
+
+        // Default construction not allowed as this leads to invalid configuration of the allocation
+        // managers.
+
+        Validator()                 = delete;
+        Validator(Validator const&) = delete;
+        Validator(Validator&&)      = default;
+        Validator& operator=(Validator const&) = delete;
+        Validator& operator=(Validator&&) = default;
+        ~Validator()                      = default;
+
+        // Validator methods
+        //
+
+        /// Validates that a module complies with the specified QIR profile.
+        bool validate(llvm::Module& module);
```

Comment:
Would be nice to clarify if this method can be called repeatedly with different modules or Validator is kind of "use once" object.
Also module isn't const, so I assume it may be transformed to bring it in compliance with the rule. Then the name of the method is a bit misleading. Also what return type means in cases when validation and transformation is performed.

Reply:
I might be wrong, but I believe that you can invoke `validate( module )` multiple times. However, it depends on the underlying LLVM routines: They only accept non-const references to the module pass manager meaning that it might or might not modify the state of the manager. I believe the only way to be sure would be to move the majority of the logic held by `Validator` into local variables in the function itself (in which case we could mark it constant).

I will update the description with the return interpretation.